ocaml/xenopsd/xc/service.ml

```diff
@@ -256,31 +258,52 @@ module DaemonMgmt (D : DAEMONPIDPATH) = struct
 
   let name = D.name
 
-  let pid ~xs domid =
+  (* Look up the commandline args for the process pid; *)
+  (* Check that they include expected args. *)
+  let is_cmdline_valid pid expected_args =
     try
-      match D.pid_location with
-      | (File file | Path_of {file; _}) when Sys.file_exists (file domid) ->
-          let path = file domid in
-          let ( let* ) = Option.bind in
-          let* pid = Unixext.pidfile_read path in
-          Unixext.with_file path [Unix.O_RDONLY] 0 (fun fd ->
-              try
-                Unix.lockf fd Unix.F_TRLOCK 0 ;
-                (* we succeeded taking the lock: original process is dead.
-                 * some other process might've reused its pid *)
-                None
-              with Unix.Unix_error (Unix.EAGAIN, _, _) ->
-                (* cannot obtain lock: process is alive *)
-                Some pid
-          )
-      | Xenstore key | Path_of {key; _} ->
-          (* backward compatibility during update installation: only has
-             xenstore pid available *)
-          let pid = xs.Xs.read (key domid) in
-          Some (int_of_string pid)
-      | _ ->
-          None
-    with _ -> None
+      let cmdline =
+        Printf.sprintf "/proc/%d/cmdline" pid
+        |> Unixext.string_of_file
+        |> Astring.String.cuts ~sep:"\000"
+      in
+      List.for_all (fun arg -> List.mem arg cmdline) expected_args
```

Comment:
Is this the right test? Don't we expect the two lists to be identical using string quality?

Reply:
The service has lot of cmdline arguments, for example for a qemu instance, please refer this:
`65568     1938  1881  0 Jan03 ?        00:10:25 qemu-dm-33 -machine pc-0.10,accel=xen,max-ram-below-4g=4026531840,allow-unassigned=true,trad_compat=True -vnc unix:/var/run/xen/vnc-33,lock-key-sync=off -monitor null -pidfile /var/run/xen/qemu-dm-33.pid -xen-domid 33 -m size=4088 -boot order=dc -usb -device usb-tablet,port=2 -smp 2,maxcpus=2 -chardev file,id=serial0,append=on,path=/tmp/local/scratch/xenrt/guest-console-logs/console.33.log -serial chardev:serial0 -display none -nodefaults -trace enable=xen_platform_log -sandbox on,obsolete=deny,elevateprivileges=allow,spawn=deny,resourcecontrol=deny -S -global PIIX4_PM.revision_id=0x1 -global ide-hd.ver=0.10.2 -global piix3-ide-xen.subvendor_id=0x5853 -global piix3-ide-xen.subsystem_id=0x0001 -global piix3-usb-uhci.subvendor_id=0x5853 -global piix3-usb-uhci.subsystem_id=0x0001 -global rtl8139.subvendor_id=0x5853 -global rtl8139.subsystem_id=0x0001 -parallel null -qmp unix:/var/run/xen/qmp-libxl-33,server,nowait -qmp unix:/var/run/xen/qmp-event-33,server,nowait -device xen-platform,addr=3,device-id=0x0001,revision=0x2,class-id=0x0100,subvendor_id=0x5853,subsystem_id=0x0001 -drive file=/dev/sm/backend/f4099437-827f-9c2f-4917-09ba8134822c/82d9cfa5-0edd-4654-b263-88f802f12167,if=none,id=ide0-hd0,auto-read-only=off,format=raw -device ide-hd,drive=ide0-hd0,bus=ide.0,unit=0,bios-chs-trans=forcelba -device rtl8139,netdev=tapnet0,mac=46:41:4a:2c:fd:f2,addr=4 -netdev tap,id=tapnet0,fd=7 -device VGA,vgamem_mb=8,addr=2,romfile=,rombar=1,subvendor_id=0x5853,subsystem_id=0x0001,qemu-extended-regs=false -vnc-clipboard-socket-fd 4 -chardev stdio,id=ovmf -device isa-debugcon,chardev=ovmf,iobase=0x402 -xen-domid-restrict -chroot /var/xen/qemu/root-33 -runas 65568:1023
qemu-dm-33-machinepc-0.10,accel=xen,max-ram-below-4g=4026531840,allow-unassigned=true,trad_compat=True-vncunix:/var/run/xen/vnc-33,lock-key-sync=off-monitornull-pidfile/var/run/xen/qemu-dm-33.pid-xen-domid33-msize=4088-bootorder=dc-usb-deviceusb-tablet,port=2-smp2,maxcpus=2-chardevfile,id=serial0,append=on,path=/tmp/local/scratch/xenrt/guest-console-logs/console.33.log-serialchardev:serial0-displaynone-nodefaults-traceenable=xen_platform_log-sandboxon,obsolete=deny,elevateprivileges=allow,spawn=deny,resourcecontrol=deny-S-globalPIIX4_PM.revision_id=0x1-globalide-hd.ver=0.10.2-globalpiix3-ide-xen.subvendor_id=0x5853-globalpiix3-ide-xen.subsystem_id=0x0001-globalpiix3-usb-uhci.subvendor_id=0x5853-globalpiix3-usb-uhci.subsystem_id=0x0001-globalrtl8139.subvendor_id=0x5853-globalrtl8139.subsystem_id=0x0001-parallelnull-qmpunix:/var/run/xen/qmp-libxl-33,server,nowait-qmpunix:/var/run/xen/qmp-event-33,server,nowait-devicexen-platform,addr=3,device-id=0x0001,revision=0x2,class-id=0x0100,subvendor_id=0x5853,subsystem_id=0x0001-drivefile=/dev/sm/backend/f4099437-827f-9c2f-4917-09ba8134822c/82d9cfa5-0edd-4654-b263-88f802f12167,if=none,id=ide0-hd0,auto-read-only=off,format=raw-deviceide-hd,drive=ide0-hd0,bus=ide.0,unit=0,bios-chs-trans=forcelba-devicertl8139,netdev=tapnet0,mac=46:41:4a:2c:fd:f2,addr=4-netdevtap,id=tapnet0,fd=7-deviceVGA,vgamem_mb=8,addr=2,romfile=,rombar=1,subvendor_id=0x5853,subsystem_id=0x0001,qemu-extended-regs=false-vnc-clipboard-socket-fd4-chardevstdio,id=ovmf-deviceisa-debugcon,chardev=ovmf,iobase=0x402-xen-domid-restrict-chroot/var/xen/qemu/root-33-runas65568:1023`

Reply:
Are the two not identical because we have some wrapper script that makes adjustment? In that case the current approach to look up a few critical command line arguments would be ok.